### PR TITLE
feat: support schema autocomplete

### LINKED ISSUE #216: Table and schema names do not autocomplete properly
Currently the editor auto completes column names, but does not auto complete schmas or tables in certain circumstances.

- Schemas should auto-complete along with table names after typing `from`, `join`, etc.
- Tables should auto-complete after typing `schema.`, same as columns auto-complete after typing `table.`

### LINKED ISSUE #976: BUG: Code Completion doesn't autocomplete schemaname or schemaname.tablename
I am connected to my database which has roughly 1200 objects and about 20 schemas.  The objects load and then are displayed on the left navigation window listing tables, functions etc.

I type out the following in the sql editor:

select * from schema.tablename;

While I am typing, my expectation is that as soon as I start typing "schema", I would be provided recommendations of schema.  It does provide recommendations for table, but I do not like setting my search_path so I like code sql with the full declaration of schema.tablename.  And after I type "schema." there are no tablename recommendations.

I like your app, but this is my main issue with postgres sql tools.  Looks good and I will continue to follow your app.

Also, when I right click on a function in the navigation window and select SQL:Create, nothing happens where I would expect it to open the DDL in a tab.

SYSTEM : Ubuntu 20.04
BEEKEEPER : 3.0.7

### LINKED ISSUE #1184: FEAT: Postgres autocomplete all schemas
Only the public schema tables get autocomplete right now, and there's no autocomplete if you type `schema.`, but it should autocomplete tables here too.

Ripped from #173

### PR DESCRIPTION
**Pull request by azmy60**
fix #976 | fix #216 | fix #1184 

This is an attempt to support schema autocomplete with table alias and columns. I replaced some logics in the vendor file as it can't be modified anywhere outside nor tested. It's far easier to solve this when I can unit test the codes :smile:. Let me know if there are things that need to be worked around. I've also tested this in postgres and sqlite (making sure it doesn't break in database without schemas).

bks-support-schema

### CODE REVIEW

**Comment by rathboma** [comment]
Wow, this is awesome! Going to take a look today. Thanks for the killer PR :-)

File: apps/studio/src/lib/editor.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        text: table.schema ? `${table.schema}.${tableName}` : tableName,
+        displayText: table.name,
+      };
+    })
+  );
+}
+
+const tableNameValidationRegex = /(?:[^a-z0-9_]|^\d)/;
+
+export function sanitizeTableName(name: string): string {
```
Comment (rathboma) [inline apps/studio/src/lib/editor.ts:102]: Can this be something added to the `DialectData` for all connection types? Then we can just pass in the dialect data for each connection.

We already have a `wrapIdentifier`, you could add this as `maybeWrapIdentifier` or `friendlyNormalizedIdentifier`? [See this file as an example](https://github.com/beekeeper-studio/beekeeper-studio/blob/e405db30297950a4295c9d1a15e78aa2713e9a45/shared/src/lib/dialects/postgresql.ts#L4)

Comment (rathboma) [reply apps/studio/src/lib/editor.ts:102]: The dialect data is available in the store as a getter, [see the getter here](https://github.com/beekeeper-studio/beekeeper-studio/blob/e405db30297950a4295c9d1a15e78aa2713e9a45/apps/studio/src/store/index.ts#L128).

So you could do something like:

```js
computed: {
 ...mapGetters('dialectData')
}

```

Then: `makeDbHint(tables, this.dialectData)`

File: apps/studio/tests/unit/lib/editor.spec.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  beforeAll(() => {
+    dbHint = makeDBHint(tableOrViews, "postgresql");
+  });
+
+  describe("makeDBHint", () => {
+    it("should make arrays of tables and schemas", () => {
+      expect(makeDBHint(tableOrViews, "postgresql")).toEqual({
+        tables: [
+          { name: "my_table", schema: "public" },
```
Comment (rathboma) [inline apps/studio/tests/unit/lib/editor.spec.js:23]: In postgres if you create a normal table using quotes, like `'"normal_table"'`, it always has to be referenced with the quotes I think? Does it work in that case?

(didn't test this, just pulling it from my memory)

File: apps/studio/tests/unit/lib/editor.spec.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  parseDBHintTable,
+  parseDBHintTables,
+  queryTable,
+  findTablesBySchema,
+  pushTablesToResult,
+  sanitizeTableName,
+} from "../../../src/lib/editor";
+import { tableOrViews } from "../../fixtures/tables";
+
+describe("lib/editor", () => {
```
Comment (rathboma) [inline apps/studio/tests/unit/lib/editor.spec.js:12]: Please correct me if I'm wrong here, but it looks like this won't work if the DB has schemas, and no schema is provided in the query?

Postgres by default uses the `public` schema, and SQL Server uses `dbo`, and not specifying this is totally legit, and really common honestly.

A good unit test would be to check that it can resolve:

- `schema.table`
- `schema.table.column`
- `table`
- `table.column`

The complex bit is that `a.b` could be either `schema.table` or `table.column`.

It gets a little more complex in that there could be both `schema: a`, and `table: public.a`, so the hint system would need to provide some ability to handle this situation. So if someone types `a.`, it should probably autocomplete both table names from `schema: a` and column names from `table: public.a`.

Default schema is [also provided in the store as a getter](https://github.com/beekeeper-studio/beekeeper-studio/blob/e405db30297950a4295c9d1a15e78aa2713e9a45/apps/studio/src/store/index.ts#L104), so should be easy to integrate if not already?

**Review by rathboma — CHANGES_REQUESTED** [review]
This is 🚀 🚀 🔥 🔥 . Really amazing work to figure this out.

I have two real pieces of feedback:

1. Can we generalize the table name sanitization?
2. Can we handle default schemas and ambiguous table/schema matches?

File: apps/studio/tests/unit/lib/editor.spec.js (hunk 3 above)
Comment (azmy60) [reply apps/studio/tests/unit/lib/editor.spec.js:12]: Oh, you've got a point. I forgot to add the default schema.

I also think this part is a little bit confusing because there are [two ways](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-8ada96660a46ee9c6784d974ac1043ced5db7c7f1be4eab25c9e25ecb43fbd3bR66-R70) to find tables right now: one by looking up the key in `tables` object, the other one is by using the `hintTables` array. I add the second one because the `tables` can overwrite the same table names that have different schemas, while the `hintTables` can not.

I'll try to add the test to clarify this one and fix the other feedbacks soon.

File: apps/studio/tests/unit/lib/editor.spec.js (hunk 2 above)
Comment (azmy60) [reply apps/studio/tests/unit/lib/editor.spec.js:23]: Yup. Table names that are case sensitive will need to be referenced with quotes. I'll add this to the test.

**Comment by azmy60** [comment]
@rathboma I've added the default schema and `maybeWrapIdentifier` to the `dialectData`. The trickiest part here is to resolve `schema.table.column` where the table uses quotes like `public."table"`. Is this good enough to cover the edge cases? I'm not entirely aware of all databases syntax.

File: apps/studio/tests/unit/lib/editor.spec.js (hunk 3 above)
Comment (azmy60) [reply apps/studio/tests/unit/lib/editor.spec.js:19]: Your test suggestions should now be covered by the [queryTable test](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-4e044bf2110531d177d19691735abb6cea1e27c92e792eed2734814d8b274ab9R144). As long as it resolves the `schema.table`, it should be able to resolve `schema.table.column` too which depends on the column lookup [here](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-8bd68c134db51e45491610ee0f614c52db0be8412bf23eded25bf6aeffa086a9R918). It should also work when schema is not provided or `undefined` [(test)](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-4e044bf2110531d177d19691735abb6cea1e27c92e792eed2734814d8b274ab9R126-R142).

**Comment by azmy60** [comment]
I just notice that if you write <code>schema.\`table\`</code> with quotation instead of `schema.table`, it won't show the autocomplete. Would you like this to be fixed as well?

**Comment by not-night-but** [comment]
@azmy60 If we support other quotations (`'` and `"`) then I think it's best we support backtick as well

**Comment by azmy60** [comment]
Ok. I'll start working on that tomorrow.

**Comment by not-night-but** [comment]
@azmy60 I'm just messing around with this a bit, and I have a thought:

I think the default behaviour (without specifying a schema) should remain the same. As it stands right now, if I don't specify a schema in postgres I get all the tables from information_schema and pg_catalog before I get any tables from public or any schemas.

Old behaviour: 
oldAutoComplete

New behaviour:
newAutoComplete

Note that in the new behaviour, choosing one of the autocomplete options from other schemas actually creates an incorrect autocompletion, as you'll get a `relation doesn't exist` error because the correct schema isn't specified. It should probably show 
1. Tables from the default schema
2. schemas
3. keywords

**Review by not-night-but — CHANGES_REQUESTED** [review]
@azmy60 I'm just messing around with this a bit, and I have a thought:

I think the default behaviour (without specifying a schema) should remain the same. As it stands right now, if I don't specify a schema in postgres I get all the tables from information_schema and pg_catalog before I get any tables from public or any schemas.

Old behaviour: 
oldAutoComplete

New behaviour:
newAutoComplete

Note that in the new behaviour, choosing one of the autocomplete options from other schemas actually creates an incorrect autocompletion, as you'll get a `relation doesn't exist` error because the correct schema isn't specified. It should probably show 
1. Tables from the default schema
2. schemas
3. keywords

File: apps/studio/src/lib/editor.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ */
+type WordList = Record<string, Word>;
+
+export interface DBHint {
+  defaultSchema?: string;
+  /**
+   * A list of tables in key-value form for fast queries, where the key
+   * is the table name. If the database supports schemas, all tables are from
+   * the default schema.
+   **/
```
Comment (not-night-but) [inline apps/studio/src/lib/editor.ts:22]: Love that you're documenting your code, thanks so much!

File: apps/studio/src/lib/editor.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+const quot = "'\"`";
+
+const SCHEMA_TABLE = new RegExp(
+  "(?:(?<schema>[quot]?.+?[quot]?)\\.(?<table1>[quot]?.+[quot]?)|(?<table2>[quot]?.+[quot]?))".replaceAll(
+    "quot",
+    quot
+  )
+);
```
Comment (not-night-but) [inline apps/studio/src/lib/editor.ts:87]: Unless you have a reason I'm not seeing here, could you replace `quot` with `quote`?

File: apps/studio/src/vendor/sql-hint/index.js [hunk 6]
```diff
@@ -231,22 +257,42 @@
         var lineText = query[i];
         eachWord(lineText, function(word) {
           var wordUpperCase = word.toUpperCase();
-          if (wordUpperCase === aliasUpperCase && getTable(previousWord))
+          if (wordUpperCase === aliasUpperCase && (tableWord = getTable(previousWord)))
```
Comment (not-night-but) [inline apps/studio/src/vendor/sql-hint/index.js:260]: If possible I would prefer if we didn't assign within conditionals. I'm also not sure that `table` is entirely necessary anymore, but I haven't dug around in this area of the code.

File: apps/studio/src/vendor/sql-hint/index.js [hunk 7]
```diff
@@ -231,22 +257,42 @@
...
+    const lineText = editor.display.view[cur.line].line.text;
+    const lineTextBeforeDot = lineText.substring(0, cur.ch - 1)
+    const wordStart = lineTextBeforeDot.lastIndexOf(" ") + 1
+    const word = lineTextBeforeDot.split(" ").pop()
+    return [word, wordStart]
+  }
+
+  function addTablesBySchema(result, search) {
+    const [schema] = splitSchemaTable(search);
+      const unwrappedSchema = dbHint.dialect.unwrapIdentifier(schema);
```
Comment (not-night-but) [inline apps/studio/src/vendor/sql-hint/index.js:282]: Indentation :stuck_out_tongue_winking_eye:

**Review by not-night-but — CHANGES_REQUESTED** [review]
@azmy60 Great job! Unfortunately I think I found another bug :face_exhaling: 

It seems if I have a quoted schema, it won't autocomplete the tables:
quotedSchemaIssue

Hopefully it's an easy fix, and then we can see what @rathboma thinks :grinning:

File: apps/studio/src/vendor/sql-hint/index.js (hunk 6 above)
Comment (azmy60) [reply apps/studio/src/vendor/sql-hint/index.js:260]: I assigned within conditionals because I thought it could minimize the lines of changes. But I'll fix that if it's preferred no to.

File: apps/studio/src/vendor/sql-hint/index.js (hunk 7 above)
Comment (azmy60) [reply apps/studio/src/vendor/sql-hint/index.js:282]: lol 😆

File: apps/studio/src/vendor/sql-hint/index.js (hunk 6 above)
Comment (not-night-but) [reply apps/studio/src/vendor/sql-hint/index.js:260]: I personally just think it makes the code a bit harder to read.

**Comment by azmy60** [comment]
@not-night-but I've added the fixes to your feedbacks. The problem was if the schema/table contains spaces like `SELECT * FROM "my schema".table`, it would parse it as `["SELECT", "*", "FROM", '"my', 'schema".table']`. It should be fixed based on [this test](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-4e044bf2110531d177d19691735abb6cea1e27c92e792eed2734814d8b274ab9R272-R277) and works for any quotation characters.

However, the autocompletion wouldn't work if the table name contains a quote like `my'table` (it's a valid name in mysql and postgres, though it's probably not a good practice 😅). This could be fixed by updating [this regex](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-ff50d40fa14d8bc2b3c86de9f768caed39751c7b3fed18958816733cd1ed409bR162) or the [splitWords()](https://github.com/beekeeper-studio/beekeeper-studio/pull/1610/files#diff-ff50d40fa14d8bc2b3c86de9f768caed39751c7b3fed18958816733cd1ed409bR164) function. But I haven't found any solution yet.

**Comment by not-night-but** [comment]
@azmy60 I'm so sorry I missed this!!! Thanks so much. I'll take a look in a bit and see if I can figure out a way to support that weird syntax. In my experience there is a lot of stuff that is supported in postgres that really shouldn't be 😂. When I found out you could put spaces in schema/table names I almost had an aneurysm.

File: apps/studio/src/lib/editor.ts [hunk 8]
```diff
@@ -77,12 +89,10 @@ export function makeDBHint(
...
-
 const SCHEMA_TABLE = new RegExp(
-  "(?:(?<schema>[quot]?.+?[quot]?)\\.(?<table1>[quot]?.+[quot]?)|(?<table2>[quot]?.+[quot]?))".replaceAll(
-    "quot",
-    quot
+  "(?:(?<schema>[quote]?.+?[quote]?)\\.(?<table1>[quote]?.+[quote]?)|(?<table2>[quote]?.+[quote]?))".replaceAll(
+    "quote",
+    "'\"`"
   )
 );
```
Comment (not-night-but) [inline apps/studio/src/lib/editor.ts]: @azmy60 As I'm looking at this again, I'm wondering what the `table2` capture group is for here. Maybe I'm just bad at regex, but it seems to me like `table1` will always be the group that matches, yes?

Comment (not-night-but) [reply apps/studio/src/lib/editor.ts]: Nevermind I'm dumb lol

**Comment by not-night-but** [comment]
@azmy60 as for fixing the regex to allow for `my'table`, I believe this should fix it:

```regex
(?:(?<schema>(['"`]?).+?\2?)\.(?<table1>(['"`]?).+\4?)|(?<table2>(['"`]?).+\6?))
```

**Comment by azmy60** [comment]
@not-night-but Thanks for the help! I just took a quick look, it turned out that the issue was not from the regex🤦. The issue was from the CodeMirror (probably the sql mode module?) that didn't [get the token](https://github.com/beekeeper-studio/beekeeper-studio/blob/94b1cee9c15397c6c0f60ca6d530728221a1bc0c/apps/studio/src/vendor/sql-hint/index.js#L312) correctly.

So If I write `"my'table".`, it gives me `'table".` as the token. The correct token should just be the `.` character. You can see that it also breaks the highlight and the autocomplete. I recorded this using the current version of Beekeeper Studio.

[screen-recorder-thu-jul-13-2023-09-54-46.webm](https://github.com/beekeeper-studio/beekeeper-studio/assets/38707148/8f2b2953-bad9-4824-9d50-f4aa87a85253)

**Comment by not-night-but** [comment]
@azmy60 hmmmm. Yeah there are a couple quirks with our highlighting system.

I would still advise that we switch the regex to something like what I sent above, as it ensures that the closing quote is the same as the opening quote

**Comment by azmy60** [comment]
@not-night-but done :)

**Comment by not-night-but** [comment]
@azmy60 Coolio, thanks! So beyond the weird highlighting issues with that one example, were there any other issues you saw?

**Comment by azmy60** [comment]
I didn't notice any other issues so far.

**Comment by not-night-but** [comment]
@azmy60 alrighty cool. I'll play around with it a bit more in a bit, and if I don't find anything I'll ask Matthew if we're good to merge 😃

**Review by not-night-but — APPROVED** [review]
